src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs

```diff
@@ -0,0 +1,895 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    [CompilerTrait(CompilerFeature.IOperation)]
+    public partial class IOperationTests : SemanticModelTestBase
+    {
+        [Fact]
+        public void CoalesceAssignment_SimpleCase()
+        {
+
+            string source = @"
+class C
+{
+    static void M(object o1, object o2)
+    {
+        /*<bind>*/o1 ??= o2/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation (OperationKind.CoalesceAssignment, Type: System.Object) (Syntax: 'o1 ??= o2')
+  Target: 
+    IParameterReferenceOperation: o1 (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'o1')
+  WhenNull: 
+    IParameterReferenceOperation: o2 (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'o2')
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<AssignmentExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact]
+        public void CoalesceAssignment_WithConversion()
+        {
+            string source = @"
+class C
+{
+    static void M(object o1, string s1)
+    {
+        /*<bind>*/o1 ??= s1/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation (OperationKind.CoalesceAssignment, Type: System.Object) (Syntax: 'o1 ??= s1')
+  Target: 
+    IParameterReferenceOperation: o1 (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'o1')
+  WhenNull: 
+    IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Object, IsImplicit) (Syntax: 's1')
+      Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: True, IsUserDefined: False) (MethodSymbol: null)
+      Operand: 
+        IParameterReferenceOperation: s1 (OperationKind.ParameterReference, Type: System.String) (Syntax: 's1')
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<AssignmentExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact]
+        public void CoalesceAssignment_NoConversion()
+        {
+            string source = @"
+class C
+{
+    static void M(C c1, string s1)
+    {
+        /*<bind>*/c1 ??= s1/*</bind>*/;
+    }
+}
+";
+            string expectedOperationTree = @"
+ICoalesceAssignmentOperation (OperationKind.CoalesceAssignment, Type: ?, IsInvalid) (Syntax: 'c1 ??= s1')
```

Comment:
>Type: ? [](start = 64, length = 7)

I think we can safely say that the type is ```C```

Reply:
This follows the same convention that other compound assignments do, where if there is no conversion from the left to the right, the whole compound expression has the error type. See `ICompoundAssignment_BinaryOperatorOutConversion_InvalidMissingInConversion`.

---
In reply to: [209401968](https://github.com/dotnet/roslyn/pull/29063#discussion_r209401968) [](ancestors = 209401968)